chore(docs): New "Creating a Source Plugin" tutorial (#37538)
gatsbyjs/gatsby: docs/docs/tutorial/creating-a-source-plugin/part-3/index.mdx

---
title: "Part 3: Define GraphQL Schema"
---

import { LinkButton } from "gatsby-interface"
import Collapsible from "@components/collapsible"
import { MdArrowForward } from "react-icons/md"

## Introduction

In the last part you learned how to create GraphQL nodes and use them in your site. In this part you'll dive deeper into the GraphQL schema creation process and learn how to modify the schema. This knowledge will be valuable for the reliability and usability of your plugin.

> One of Gatsby’s main strengths is the ability to query data from a variety of sources in a uniform way with GraphQL. For this to work, a GraphQL schema must be generated that defines the shape of the data. Gatsby is able to automatically infer a GraphQL Schema from your data, and in many cases, this is really all you need. There are however situations when you either want to explicitly define the data shape, or add custom functionality to the query layer — this is what Gatsby’s `createSchemaCustomization` Node API provides.

Reviewer: Small suggestion to change for readability:
```suggestion
One of Gatsby’s main strengths is the ability to query data from a variety of sources in a uniform way with GraphQL. For this to work, a GraphQL schema must be generated that defines the shape of the data. Gatsby is able to automatically infer a GraphQL Schema from your data, and in many cases, this is really all you need. However, there are situations when you either want to explicitly define the data shape, or add custom functionality to the query layer — this is what Gatsby’s `createSchemaCustomization` Node API provides.
```

In this part of the tutorial, you'll learn how to use the [`createSchemaCustomization` Node API](/docs/reference/config-files/gatsby-node/#createSchemaCustomization).

By the end of this part of the tutorial, you will be able to:

- Understand how Gatsby automatically creates GraphQL types
- Explicitly define the GraphQL schema for your plugin
- Create a foreign-key relationship between your `Post` and `Author` nodes

## Automatic type inference

> Before explaining how you can explicitly define your GraphQL schema, it's important for you to understand how Gatsby's automatic type inference works. It'll then make more sense why you'd want to define a schema manually and what benefits you'll receive.

Reviewer: Suggestion to change for readability:
```suggestion
Before explaining how you can explicitly define your GraphQL schema, it's important for you to understand how Gatsby's automatic type inference works. Then, it'll make more sense why you'd want to define a schema manually and what benefits this offers.
```

export { onPluginInit } from "./on-plugin-init"
export { sourceNodes } from "./source-nodes"
// highlight-next-line
export { createSchemaCustomization } from "./create-schema-customization"
```

Also double check that you're running the `develop:deps` script in your terminal so that your changes to the plugin are compiled.

### `createTypes`

Before using [`createTypes`](/docs/reference/config-files/actions/#createTypes), here are some things to note for its usage:

- Type definitions can be provided either in GraphQL's [schema definition language (SDL)](https://graphql.org/learn/schema/), Gatsby Type Builders, or a combination of both. You'll use the SDL syntax for the rest of this tutorial but you can also read an example for Gatsby's Type Builder syntax in this part.
- By default, explicit type definitions you add with `createTypes` will be merged with inferred field types. So if you only want to define a subset of fields, the rest will be inferred as before. You can modify this behavior with the `@infer` and `@dontInfer` extensions. You'll learn more about extensions in [Add a foreign-key relationship](#add-a-foreign-key-relationship).

> - Type definitions targeting root node types, e.g. `MarkdownRemark` or others added in `sourceNodes`/`onCreateNode` like `Post` and `Author`, need to implement the `Node` interface. You can do this by adding the `implements Node` to the SDL.

Reviewer: Possible typo?
```suggestion
- Type definitions targeting root node types, e.g. `MarkdownRemark` or others added in `sourceNodes`/`onCreateNode` like `Post` and `Author`, need to implement the `Node` interface. You can do this by adding `implements Node` to the SDL.
```

```json
{
  "name": "Gatsby",
  "address": {
    "street": "Some street",
    "city": "San Francisco",
    "state": "California",
  }
}
```

In this example, let's pretend that you know `state` is not always provided — meaning that you want to explicitly define it to avoid the situation described above. Conversely, you also know that `street` will always be defined, and as such, you can mark it as [Non-Null](https://graphql.org/learn/schema/#lists-and-non-null) with a `!`. Lastly, `city` will be automatically inferred and merged with the custom types.

```ts
import type { GatsbyNode } from "gatsby"

export const createSchemaCustomization: GatsbyNode[`createSchemaCustomization`] =
  ({ actions }) => {
    const { createTypes } = actions

    createTypes(`
      type User implements Node {
        address: UserAddress
      }

      type UserAddress {
        street: String!
        state: String
      }
    `)
  }
```

In this example the `User` type is a root node type, so you'll be able to query it with `allUser` and `user` in GraphQL.

<Collapsible
  summary={<em>Gatsby Type Builder example</em>}
>

> In many cases, GraphQL SDL provides a succinct way to provide type definitions for your schema. If however you need more flexibility, `createTypes` also accepts type definitions provided with the help of Gatsby Type Builders, which are more flexible than SDL syntax. They are accessible on the `schema` argument passed to Node APIs.

Reviewer 2: Not necesairly need to change anything, but they way I view SDL vs type builders is like this:

SDL is more terse, easier to read and to understand. Best choice for hardcoded schemas unless you need custom resolvers which force you to do something on top of SDL to provide those resolvers (either register custom directive or do both SDL and type builder / createResolvers). It becomes less easy to read if SDL is being generated through some kind of string concatenation/interpolation/manipulation.

Type Builders are better suited for cases when you can't "hardcode" schema and/or you need custom resolvers - if CMS has ability to create custom types with custom fields and you can fetch those definitions - it's easier to use typebuilders than generating SDL that would match it

This is a really powerful ability in Gatsby's data layer.

Imagine this example: "Team Gatsby" likes to author their blog posts in Markdown, "Team Daisy" is responsible for keeping author information up do date and keeps track of that in a CMS. As long as both teams use the same unique identifier, both nodes can be linked together in Gatsby's data layer. Both teams can work independently on their data.

</Announcement>

In the previous section you already defined the GraphQL types for `Post` and `Author` so all the necessary boilerplate is already done. You can now use the `@link` extension (as a directive in the SDL). Before showing that, here's a short explanation on the `@link` extension:

- `@link` can receive two arguments (`by` and `from`) but if no argument is given, Gatsby will use the `id` field as the foreign-key field (equivalent to `@link(by: "id")`). You'll use this default behavior in [Part 6](/docs/tutorial/creating-a-source-plugin/part-6/) for the Image CDN feature.
- `by` is the field you link **to** (on the target node).

> - `from` is the field you link **from** (on the current node).

Reviewer 2: There is usage of `from` here, so I wonder it warrants a note saying that you would want to use that argument if you want different name for field in graphql that what Node has named the value?

So as example if node has `author_id` field, you wouldn't want field like that in schema. Instead you want `author` field that uses current's node `author_id` for source of value to match target node on?

Feel free to ignore (or maybe I'll discard it if there is mention of it later on)

      <section className="posts-grid">
        {posts.map((post) => (
          <Link key={post.id} to={post.slug} className="posts-card">
            <h2>{post.title}</h2>
            {/* highlight-next-line */}
            <span>Author: {post.author.name}</span>
          </Link>
        ))}
      </section>
    </main>
  )
}

// Head API

export const query = graphql`
  query IndexPage {
    allPost {
      nodes {
        id
        slug
        title
// highlight-start
        author {
          name
        }
// highlight-end
      }
    }
  }
`
```

And in `src/pages/{Post.slug}.tsx` like so:

```tsx:title=site/src/pages/{Post.slug}.tsx
// Imports

export default function PostPage({
  data: { post },
}: PageProps<{ post: Queries.Post }>): React.ReactElement {
  return (
    <main>
      <h1>{post.title}</h1>
      {/* highlight-next-line */}
      <p>Author: {post.author.name}</p>
      <br />
      <Link to="/">Back to home page</Link>
    </main>
  )
}

// Head API

export const query = graphql`
  query PostPage($slug: String!) {
    post(slug: { eq: $slug }) {
      title
// highlight-start
      author {
        name
      }
// highlight-end
    }
  }
`
```

> Go visit your index page in the browser and everything should be working again.

Reviewer: Suggestion for brevity:
```suggestion
Visit your index page in the browser and everything should be working again.
```
